### ISSUE TITLE
Create a way to review developer changes without affecting the production build

### ISSUE THREAD

**Issue by finanalyst** [Raku/doc-website#217]
We have some UI changes (eg new search) that need to be tested with whole site (all the Rakudoc sources on `Raku/doc`) on browsers, but without messing up the production build of `docs.raku.org`.

I propose:
- creating a new `Mode` (structurally just a sub-directory of `doc-website/`) called **Development**. This would clone **Website**.
- Changes pushed to **Development** would trigger a rebuild and deployment of `docs-dev.raku.org`
- **Website** (the production build) would thus be left unchanged, and so pushes to `Raku/doc` could trigger a rebuild of `docs.raku.org` automatically.
- This would allow for more lenient reviews of PRs of changes to **Development** and much stricter review standards for **Website**

I can quickly set up the new Mode on `Raku/doc-website` but I don't know how to get changes to **Development** to show up on `docs-dev.raku.org`.

UI changes could include new CSS, new functionality, new plugins, new templates. A review for **Website** would include approving the effect of suggested changes on **Development**.

@dontlaugh @coke @2colours @cfa any feedback on this?

**Comment by 2colours** [comment]
I don't fully get what a new mode would solve that version control and CI tooling cannot. From what I know, `docs-dev.raku.org` is built regularly, while `docs.raku.org` deployment is already kind of a "push the red button" thing.

Also, I have the impression that using `docs-dev.raku.org` as a sort of CI playground is not manageable if multiple people propose rather divergent changes. Ultimately, we need to be able to have builds for several variations for the site; either a common "farm", or tooling to set up developer environments and build ourselves.

**Comment by coke** [comment]
We need two pipelines: one for anything on main branch for both repos, and one that can be tied to *any* branch on either repo and built/pushed as needed. This allows us to test more risky/invasive changes without blocking simple doc updates.

If this turns out to also be fronted by docs-dev, we need a way to switch them back and forth.- need to keep regular site updates available. And yes, we need to coordinate which single set of changes is being tested at any one time. It's not a free for all.

**Comment by dontlaugh** [comment]
I have some time today to work on this issue. 

> Also, I have the impression that using docs-dev.raku.org as a sort of CI playground is not manageable if multiple people propose rather divergent changes

I agree with this. I think docs-dev.raku.org should effectively track the `main` branch, and we should strive to keep it in a _potentially_ deployable state 24/7

> We need two pipelines: one for anything on main branch for both repos, and one that can be tied to any branch on either repo and built/pushed as needed.

I also agree with this. 

----

Here is the architecture I am aiming for.

![Google Draw](https://docs.google.com/drawings/d/e/2PACX-1vTz1M_gFK6bKlggYUF3Y2tCmyV-8beh1SF-mDWIHXSUptV_JUNcI7oZb4zNp2j-UTk5Vbz2yefaAxhP/pub?w=1793&h=959)

It should be possible to serve all of our websites off of two identical servers, with a single software LB in front. This works as long as:

1. The websites are truly stateless (Raku/doc-website is)
2. We run an additional reverse proxy on each server
3. The reverse proxy is configurable via automation (Caddy has an API)
4. We deploy agents (such as a buildkite-agent) to handle deploys, and other utilities and scripts to support

The Raku/Perl Foundation has just recently created a Hetzner space for me to migrate our existing deployment into. I'm going to take this opportunity to continue writing up a plan around this.

So, long story short, the goal is to serve dev stuff off of a subdomain, probably under `*.docs-dev.raku.org`. And to avoid opening a port for every little website, we need to deploy a standalone reverse-proxy.

And the pattern of using two identical servers to deploy everything will require re-architecting our current setup (which uses a server for dev and a different server for prod), but if we can solve for this use case, it will scale to other projects, as long as they're stateless workloads.

**Comment by finanalyst** [comment]
This seems a much better way to handle the issue than the way I had suggested.
Just one question: the search function needs changing. Part of the search would contain a websocket that gets data from a Cro App in the server container. The data store would be generated from a static file.
The websocket origin in the javascript would be generated from `window.location.hostname` (or `...host` to get the port), so it would point back to the domain (eg `PR111.docs-dev.raku.org`) where the html file for the page is generated from. 
@dontlaugh  I think this would still fit within  the structure you just outlined. But the websocket request may need to be dealt with separately.

**Comment by dontlaugh** [comment]
> Just one question: the search function needs changing. Part of the search would contain a websocket that gets data from a Cro App in the server container. 

You'll need to find a way to do this without Cro. It's possible to statically compile search indicies that are accessible from JavaScript, so look into that.

**Comment by finanalyst** [comment]
@dontlaugh Thats the way search works at present. But the data is 6000 lines long, and it is delivered to each page. The static asset is `search-bar.js`. It is the longest to load.

Is it possible to create two containers? One for the html served by NGINX and one for a websocket service?

**Comment by lizmat** [comment]
What would be the reason for 2 containers?  OOC, couldn't that be done from a single container?

**Comment by dontlaugh** [comment]
> Thats the way search works at present.

Let's keep it that way for now. Do not underestimate the complexity of crossing the Rubicon from static-site to dynamic webserver. I haven't seen any bugs that warrant that.

**Comment by finanalyst** [comment]
@dontlaugh look at all the issues marked SEARCH. There is a considerable number of them. The search needs changing. One request is to make the data from `X<>` markup available, which would increase the data for the search by 50%. 

My idea is not to make the site dynamic per se, but for there to be a single micro-service. The service receives a short string from a Websocket in the browser, and returns the data that matches the string.

**Comment by 2colours** [comment]
Yes, keeping the fully-blown search completely client-side is part of the reason the UX is the way it is now. I wouldn't say it is the main reason but still, it's not pulled off from someone's finger that the search shouldn't be hardwired into each and every site load and triggered event...

By the way... the search operates on static data, it is aggressively cache-able since the same input would always hit the same output. What makes it so different from the web server of a static site? It absolutely doesn't need to use Cro in particular, it can use anything that can generate the same response to the same request based on a search logic.

**Comment by finanalyst** [comment]
@dontlaugh there is a discussion thread about search #127. In it I have proposed a strategy to obtain data for an extended search.

**Comment by finanalyst** [comment]
closing as resolved by tagged docs-dev branch
